File: perk/state/cache.py
"""``.perk/workflow/`` cache-tier I/O (contracts.md §8.1).

Free functions over an explicit repo ``root``. **Both
planes read and write the same files**; the TS twin is ``extension/substrate/cache.ts``. These are
state-tiering *primitives* — no workflow semantics (no ``pending-learn`` meaning, no GC
policy; the GC *policy* lives in ``perk/state/gc.py``, surfaced as ``perk state prune`` + the
``cache-gc`` doctor check). LBYL throughout; absence is a normal,
branchable condition (reads return ``None``, not an exception).
"""

import json
from pathlib import Path
from typing import Any

from pydantic import ConfigDict

from perk.boundary import StrictBoundaryModel, StrTuple, translate_validation_errors
from perk.substrate.output import user_output

# The canonical `.perk/workflow/` subtrees (public so `perk doctor` can verify the layout).
SUBDIRS: tuple[str, ...] = ("plans", "scratch/runs", "handoff", "markers")


class CacheError(ValueError):
    """A durable ``.perk/workflow/`` cache file failed schema validation.

    Subclasses ``ValueError`` (documented, deliberate) so the best-effort
    ``except (OSError, ValueError)`` cache readers across the codebase keep their
    fail-soft behavior (a malformed cache never blocks a save) without edits, and
    so it sits alongside ``json.JSONDecodeError`` (itself a ``ValueError``).
    """


class HandoffCache(StrictBoundaryModel):
    """The pre-session CLI->extension handoff blob (``handoff/<run_id>.json``).

    Open-ended (``extra="allow"``): the handoff carries arbitrary ``handoff_extra``
    keys (and whatever ``perk state new-run --handoff`` writes), recovered downstream
    by ``plan-save``/``objective create``. The known keys below are declared so those
    recovery sites get typed attribute access; undeclared keys still round-trip.
    """

    model_config = ConfigDict(frozen=True, extra="allow", strict=True)
    run_id: str
    consumed: bool
    mode: str | None = None
    stage: str | None = None
    pi_session_id: str | None = None
    objective_id: str | None = None
    node_id: str | None = None
    adopt_from: str | None = None
    supersedes: str | None = None
    consumed_learn: StrTuple = ()


class DispatchCache(StrictBoundaryModel):
    """The durable ``run_id -> plan`` dispatch record (``scratch/runs/*/dispatch.json``).

    ``plan_ref``/``run_handle`` stay opaque dicts here — their nested shape is owned by the
    ``DispatchRecord``/``RunHandle`` domain models, not validated at this tier.
    """

    run_id: str
    stage: str
    plan_ref: dict[str, Any]
    runner: str
    kind: str
    status: str
    dispatched_at: str
    run_handle: dict[str, Any] | None = None
    error: str | None = None


class PlanRefCache(StrictBoundaryModel):
    """The active plan->branch ref pointer (``plan-ref.json``, §8.4)."""

    provider: str
    pr_id: str
    url: str
    # StrTuple (not a strict ``tuple[str, ...]``): the on-disk shape is a JSON list.
    labels: StrTuple
    objective_id: str | None = None
    consumed_learn: StrTuple = ()
    base: str | None = None


class AgentSessionCache(StrictBoundaryModel):
    """The Linear ``AgentSession`` pointer for this worktree (``agent-session.json``, §8.22)."""

    session_id: str
    issue: str
    url: str | None = None


# The land->learn semaphore: `land` sets it, `learn` clears it; while present it
# signals the worktree is not yet releasable. Single source of the name across planes (the TS
# twin is `PENDING_LEARN` in extension/substrate/cache.ts).
PENDING_LEARN = "pending-learn"


def workflow_dir(root: Path) -> Path:
    """The ``.perk/workflow/`` directory under ``root``."""
    return root / ".perk" / "workflow"


def ensure_layout(root: Path) -> Path:
    """Idempotently create the four ``.perk/workflow/`` subtrees; return the dir."""
    wd = workflow_dir(root)
    for sub in SUBDIRS:
        (wd / sub).mkdir(parents=True, exist_ok=True)
    return wd


# --- scratch: per-run inter-process files (scratch/runs/<run_id>/<name>) ---------------
#
# This module is the EXTERIOR accessor seam for scratch/session-data paths (contracts.md §8.1):
# production code never hand-builds the `scratch`/`runs` path segments
# outside this module (guard-tested by tests/test_cache_guard.py; the interior twins are
# extension/substrate/cache.ts + extension/substrate/sessionData.ts).


def scratch_dir(root: Path) -> Path:
    """The ``.perk/workflow/scratch/`` directory under ``root``."""
    return workflow_dir(root) / "scratch"


def run_scratch_dir(root: Path, run_id: str) -> Path:
    """The scratch directory for a run."""
    return scratch_dir(root) / "runs" / run_id


def list_run_ids(root: Path) -> list[str]:
    """Names of all run scratch dirs under ``scratch/runs/`` (sorted; ``[]`` when absent).

    Twin of the TS ``listRunIds`` (extension/substrate/cache.ts). Stray non-directory entries
    are ignored.
    """
    runs_root = scratch_dir(root) / "runs"
    if not runs_root.is_dir():
        return []
    return sorted(p.name for p in runs_root.iterdir() if p.is_dir())


def write_scratch(root: Path, run_id: str, name: str, content: str) -> Path:
    """Write a scratch file for a run (creating the run dir); return its path."""
    directory = run_scratch_dir(root, run_id)
    directory.mkdir(parents=True, exist_ok=True)
    path = directory / name
    path.write_text(content, encoding="utf-8")
    return path


def read_scratch(root: Path, run_id: str, name: str) -> str | None:
    """Read a scratch file, or ``None`` if it does not exist."""
    path = run_scratch_dir(root, run_id) / name
    if not path.is_file():
        return None
    return path.read_text(encoding="utf-8")


# --- session data: run-scoped session artifacts (scratch/runs/<run_id>/data/<name>) ------
#
# The session data dir — a dedicated `data/` subdir so run-scoped
# session artifacts never overlap perk machine records (dispatch.json, events.ndjson, ci-*.md)
# living directly in the run dir. Created lazily on first write; helpers degrade gracefully
# (absence and I/O failure -> ``None`` + a stderr warning, never an exception).


def session_data_dir(root: Path, run_id: str) -> Path:
    """The session data dir for a run (pure path; created lazily by ``write_session_data``)."""
    return run_scratch_dir(root, run_id) / "data"


def write_session_data(root: Path, run_id: str, name: str, content: str) -> Path | None:
    """Write a session-data file (creating the dir); return its path, or ``None`` on failure.

    Never raises: an ``OSError`` is reported loudly to stderr and yields ``None`` (a broken
    disk must not wedge a session).
    """
    directory = session_data_dir(root, run_id)
    path = directory / name
    try:
        directory.mkdir(parents=True, exist_ok=True)
        path.write_text(content, encoding="utf-8")
    except OSError as exc:
        user_output(f"warning: could not write session data {path}: {exc}")
        return None
    return path


def read_session_data(root: Path, run_id: str, name: str) -> str | None:
    """Read a session-data file; ``None`` when absent (normal, branchable) or unreadable.

    Absence is silent; an OS/decode error warns to stderr and yields ``None`` (never raises).
    """
    path = session_data_dir(root, run_id) / name
    if not path.is_file():
        return None
    try:
        return path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as exc:
        user_output(f"warning: could not read session data {path}: {exc}")
        return None


# --- handoff: pre-session CLI->extension cold-door blob (handoff/<run_id>.json) ---------


def handoff_path(root: Path, run_id: str) -> Path:
    """The handoff file path for a run."""
    return workflow_dir(root) / "handoff" / f"{run_id}.json"


def list_handoff_run_ids(root: Path) -> list[str]:
    """Stems of all handoff blobs under ``handoff/`` (sorted; ``[]`` when absent).

    Mirrors :func:`list_run_ids` so ``perk/state/gc.py`` collects orphan handoffs (no run dir) too.
    Exterior-only — there is no TS twin (GC is CLI-owned).
    """
    handoff_root = workflow_dir(root) / "handoff"
    if not handoff_root.is_dir():
        return []
    return sorted(p.stem for p in handoff_root.glob("*.json") if p.is_file())


def write_handoff(root: Path, run_id: str, data: dict[str, Any]) -> Path:
    """Write a fresh (un-consumed) handoff blob for a run; return its path.

    ``run_id`` and ``consumed: False`` are authoritative (they override anything in
    ``data``). The extension marks it consumed on claim (§8.2).
    """
    path = handoff_path(root, run_id)
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {**data, "run_id": run_id, "consumed": False}
    with translate_validation_errors(CacheError, source=str(path)):
        model = HandoffCache.model_validate(payload)
    path.write_text(
        json.dumps(model.model_dump(mode="json", exclude_unset=True), indent=2) + "\n",
        encoding="utf-8",
    )
    return path


def read_handoff(root: Path, run_id: str) -> HandoffCache | None:
    """Read + validate a handoff blob, or ``None`` if it does not exist."""
    path = handoff_path(root, run_id)
    if not path.is_file():
        return None
    with translate_validation_errors(CacheError, source=str(path)):
        return HandoffCache.model_validate(json.loads(path.read_text(encoding="utf-8")))


def mark_handoff_consumed(root: Path, run_id: str, *, pi_session_id: str | None = None) -> None:
    """Mark a handoff consumed (idempotent); a no-op if the handoff is absent.

    Keeps the file (audit + GC signal); never deletes it (the GC pitfall, §8.1).
    """
    data = read_handoff(root, run_id)
    if data is None:
        return
    update: dict[str, Any] = {"consumed": True}
    if pi_session_id is not None:
        update["pi_session_id"] = pi_session_id
    updated = data.model_copy(update=update)
    handoff_path(root, run_id).write_text(
        json.dumps(updated.model_dump(mode="json", exclude_unset=True), indent=2) + "\n",
        encoding="utf-8",
    )


# --- dispatch: the durable run_id->plan linkage for a remote drive (§8.13) -----


def dispatch_path(root: Path, run_id: str) -> Path:
    """The dispatch-record path for a remote drive (under the run's scratch dir)."""
    return run_scratch_dir(root, run_id) / "dispatch.json"


def write_dispatch(root: Path, run_id: str, data: dict[str, Any]) -> Path:
    """Write the durable ``run_id -> plan`` dispatch record (creating the run dir); return its
    path. ``run_id`` is authoritative (it overrides anything in ``data``), mirroring
    ``write_handoff``. The supervisor enumerates ``scratch/runs/*/dispatch.json``
    to correlate ``run_id <-> plan <-> PR`` (§8.13).
    """
    directory = run_scratch_dir(root, run_id)
    directory.mkdir(parents=True, exist_ok=True)
    path = directory / "dispatch.json"
    payload = {**data, "run_id": run_id}
    with translate_validation_errors(CacheError, source=str(path)):
        model = DispatchCache.model_validate(payload)
    path.write_text(
        json.dumps(model.model_dump(mode="json", exclude_unset=True), indent=2) + "\n",
        encoding="utf-8",
    )
    return path


def read_dispatch(root: Path, run_id: str) -> DispatchCache | None:
    """Read + validate a dispatch record, or ``None`` if it does not exist."""
    path = dispatch_path(root, run_id)
    if not path.is_file():
        return None
    with translate_validation_errors(CacheError, source=str(path)):
        return DispatchCache.model_validate(json.loads(path.read_text(encoding="utf-8")))


def list_dispatch_records(root: Path) -> list[DispatchCache]:
    """All dispatch records under ``scratch/runs/*/dispatch.json``, newest-first.

    Reads every run's ``dispatch.json``; skips a missing/unparseable file (loud-but-non-fatal
    to stderr, never raises — a corrupt record must not break the supervisor read). Sorted by
    ``dispatched_at`` descending (empty/missing timestamps sort last). An absent ``scratch/runs/``
    dir is the normal pre-dispatch state and yields ``[]``.
    """
    records: list[DispatchCache] = []
    for run_id in list_run_ids(root):
        path = run_scratch_dir(root, run_id) / "dispatch.json"
        if not path.is_file():
            continue
        try:
            with translate_validation_errors(CacheError, source=str(path)):
                records.append(
                    DispatchCache.model_validate(json.loads(path.read_text(encoding="utf-8")))
                )
        except (OSError, json.JSONDecodeError, CacheError) as exc:
            user_output(f"warning: skipping unreadable dispatch record {path}: {exc}")
            continue
    records.sort(key=lambda d: d.dispatched_at, reverse=True)
    return records


# --- plan-ref: the active plan->branch ref pointer (plan-ref.json) -----------------------


def plan_ref_path(root: Path) -> Path:
    """The ``cache.plan-ref`` pointer file (the local mirror of the canonical GitHub plan)."""
    return workflow_dir(root) / "plan-ref.json"


def write_plan_ref(root: Path, data: dict[str, Any]) -> Path:
    """Write the provider-agnostic plan ref (§8.4) to ``plan-ref.json``; return its path.

    ``exclude_unset`` preserves the input key set verbatim (the full real 7-key shape and the
    synthetic partial shape alike) — the on-disk shape is unchanged.
    """
    path = plan_ref_path(root)
    path.parent.mkdir(parents=True, exist_ok=True)
    with translate_validation_errors(CacheError, source=str(path)):
        model = PlanRefCache.model_validate(data)
    path.write_text(
        json.dumps(model.model_dump(mode="json", exclude_unset=True), indent=2) + "\n",
        encoding="utf-8",
    )
    return path


def read_plan_ref(root: Path) -> PlanRefCache | None:
    """Read + validate the plan ref, or ``None`` if it does not exist."""
    path = plan_ref_path(root)
    if not path.is_file():
        return None
    with translate_validation_errors(CacheError, source=str(path)):
        return PlanRefCache.model_validate(json.loads(path.read_text(encoding="utf-8")))


def plan_body_path(root: Path) -> Path:
    """The ``cache.plan`` materialized plan-body file (twin of the TS ``planBodyPath``)."""
    return workflow_dir(root) / "plan.md"


def write_plan_body(root: Path, body: str) -> Path:
    """Materialize the plan body markdown to ``plan.md`` (so in-session checkpoints can seed from
    its ``## Steps`` list); return its path."""
    path = plan_body_path(root)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(body if body.endswith("\n") else body + "\n", encoding="utf-8")
    return path


# --- agent-session: the Linear AgentSession pointer for this worktree (agent-session.json) ---


def agent_session_path(root: Path) -> Path:
    """The ``cache.agent-session`` pointer file (the Linear ``AgentSession`` mirror, §8.22)."""
    return workflow_dir(root) / "agent-session.json"


def write_agent_session(root: Path, data: dict[str, Any]) -> Path:
    """Write the Linear agent-session pointer (§8.22) to ``agent-session.json``; return its path.

    Shape: ``{"session_id": str, "issue": str, "url": str | None}`` — mirrors the
    ``write_plan_ref``/``read_plan_ref`` conventions.
    """
    path = agent_session_path(root)
    path.parent.mkdir(parents=True, exist_ok=True)
    with translate_validation_errors(CacheError, source=str(path)):
        model = AgentSessionCache.model_validate(data)
    path.write_text(
        json.dumps(model.model_dump(mode="json", exclude_unset=True), indent=2) + "\n",
        encoding="utf-8",
    )
    return path


def read_agent_session(root: Path) -> AgentSessionCache | None:
    """Read + validate the agent-session pointer, or ``None`` if it does not exist."""
    path = agent_session_path(root)
    if not path.is_file():
        return None
    with translate_validation_errors(CacheError, source=str(path)):
        return AgentSessionCache.model_validate(json.loads(path.read_text(encoding="utf-8")))


# --- markers: existence-based friction semaphores (markers/<name>) ----------------------


def marker_path(root: Path, name: str) -> Path:
    """The path of an existence-only marker file."""
    return workflow_dir(root) / "markers" / name


def set_marker(root: Path, name: str) -> Path:
    """Create an (empty) marker file; return its path."""
    path = marker_path(root, name)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.touch()
    return path


def has_marker(root: Path, name: str) -> bool:
    """True when the marker exists."""
    return marker_path(root, name).is_file()


def clear_marker(root: Path, name: str) -> None:
    """Remove a marker if present (idempotent)."""
    marker_path(root, name).unlink(missing_ok=True)

File: extension/substrate/cache.ts
// `.perk/workflow/` cache-tier I/O — the TS twin of perk/state/cache.py (contracts.md §8.1).
//
// Both planes read and write the SAME files; the cross-plane contract is the *files*, not a
// shared module. State-tiering primitives only — no workflow semantics. Imports use no
// relative paths (only node builtins), so this module loads cleanly under `node --test`.
//
// Readers are TOTAL: a corrupt/unreadable file is reported loudly on stderr (`console.error` —
// the report() seam is intentionally unavailable here, and stderr is headless-safe) and treated
// as absent (`null`), so a bad cache blob can never crash a caller mid-`session_start` before
// the read-only gate engages. The Python twins (src/perk/state/cache.py) deliberately keep
// RAISING `CacheError` (exterior plane, launch-time fail-loud) — the cross-plane contract is
// the *files*, not error semantics.
//
// Write discipline (contracts.md §8.1): every `.perk/workflow/` write goes through
// `atomicWriteFileSync` (temp file in the same directory + atomic rename) so a concurrent
// writer can never tear a file — a reader sees either the old bytes or the new bytes, never a
// mix (guard-tested by writeGuard.test.ts). The exemptions are the append-only NDJSON streams
// — the worker's `events.ndjson` (worker/stageExecution.ts) and the §8.58 hunk-watch `outbox.ndjson` /
// `delivered.ndjson` (hunkFeedback/perkFeedback.ts / hunkFeedback/store.ts) — where O_APPEND
// appends cannot truncate-tear and whole-file replace would introduce a read-modify-write race
// between independent processes. Atomicity is not mutual exclusion — whole-file
// last-writer-wins between concurrent writers is the accepted residual.

import { randomBytes } from "node:crypto";
import {
  chmodSync,
  existsSync,
  lstatSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  realpathSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { join, relative } from "node:path";

/**
 * Atomically replace `path` with `content` (the interior atomic-write seam).
 *
 * Writes a temp file in the same directory (so the rename never crosses filesystems) then swaps
 * it into place; a concurrent reader sees either the old bytes or the new bytes, never a torn
 * mix. On failure the temp file is best-effort removed and the error re-thrown. Precondition:
 * the parent directory exists (same contract as `writeFileSync`; call sites mkdir first).
 * Deliberately no fsync — crash durability is out of scope; the target is inter-process tearing
 * of regenerable, gitignored workflow state.
 */
export function atomicWriteFileSync(path: string, content: string): void {
  const tmp = `${path}.${process.pid}.${randomBytes(4).toString("hex")}.tmp`;
  try {
    writeFileSync(tmp, content, "utf8");
    renameSync(tmp, path);
  } catch (error) {
    try {
      rmSync(tmp, { force: true });
    } catch {
      // best-effort cleanup — a cleanup failure must never mask the original write/rename error
    }
    throw error;
  }
}

export interface Handoff {
  run_id: string;
  consumed: boolean;
  mode?: string;
  /** The registry stage id the cold launch primed (e.g. `objective-author`). */
  stage?: string;
  pi_session_id?: string;
  [key: string]: unknown;
}

export function workflowDir(cwd: string): string {
  return join(cwd, ".perk", "workflow");
}

/**
 * Read + parse a JSON cache blob, totally: a missing file is a silent `null` (absence is the
 * normal state); an unreadable/corrupt/wrong-shape file is a LOUD `null` (one stderr line naming
 * the file kind + path + error) — treated as absent by every caller.
 */
function readJsonOrNull<T>(path: string, what: string): T | null {
  if (!existsSync(path)) return null;
  try {
    const parsed: unknown = JSON.parse(readFileSync(path, "utf8"));
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
      throw new Error(`expected a JSON object, got ${JSON.stringify(parsed)}`);
    }
    return parsed as T;
  } catch (error) {
    console.error(`perk: unreadable ${what} at ${path} — treating as absent (${error})`);
    return null;
  }
}

// --- handoff -----------------------------------------------------------------------------

export function handoffPath(cwd: string, runId: string): string {
  return join(workflowDir(cwd), "handoff", `${runId}.json`);
}

export function readHandoff(cwd: string, runId: string): Handoff | null {
  return readJsonOrNull<Handoff>(handoffPath(cwd, runId), "handoff");
}

/** Mark a handoff consumed (idempotent); a no-op when absent. Keeps the file (audit + GC). */
export function markHandoffConsumed(
  cwd: string,
  runId: string,
  opts: { piSessionId?: string } = {},
): void {
  const data = readHandoff(cwd, runId);
  if (data === null) return;
  data.consumed = true;
  if (opts.piSessionId !== undefined) data.pi_session_id = opts.piSessionId;
  atomicWriteFileSync(handoffPath(cwd, runId), `${JSON.stringify(data, null, 2)}\n`);
}

// --- scratch -----------------------------------------------------------------------------
//
// This module is the INTERIOR path-primitive seam for scratch/session-data (contracts.md §8.1):
// production code never hand-builds the `scratch`/`runs` path segments
// outside this module (guard-tested by cacheGuard.test.ts; the ctx-level current-run seam is
// sessionData.ts and the exterior twin is perk/state/cache.py).

export function scratchDir(cwd: string): string {
  return join(workflowDir(cwd), "scratch");
}

export function runScratchDir(cwd: string, runId: string): string {
  return join(scratchDir(cwd), "runs", runId);
}

/** The run-owned directory for disposable, non-authoritative model intermediates. */
export function agentScratchDir(cwd: string, runId: string): string {
  return join(runScratchDir(cwd, runId), "agent");
}

/**
 * The session data dir for a run — a dedicated `data/` subdir so
 * run-scoped session artifacts never overlap perk machine records (dispatch.json,
 * events.ndjson, ci-*.md) living directly in the run dir. Pure path — created lazily by the
 * sessionData.ts write helpers.
 */
export function sessionDataDir(cwd: string, runId: string): string {
  return join(runScratchDir(cwd, runId), "data");
}

/** Reject run ids that could select anything except one child of the shared runs directory. */
function assertSafeRunId(runId: string): void {
  if (
    runId.length === 0 ||
    runId === "." ||
    runId === ".." ||
    runId.includes("/") ||
    runId.includes("\\") ||
    runId.includes("\0")
  ) {
    throw new Error(`refusing unsafe run id ${JSON.stringify(runId)}`);
  }
}

/**
 * Ensure one checkout-owned path component is a real directory, never a static redirect.
 * Check-before-create races against a same-UID process are intentionally out of scope.
 */
function ensureUnredirectedDirectory(
  path: string,
  opts: { createMode: number; rejectGroupWorldWrite: boolean },
): void {
  let stat: ReturnType<typeof lstatSync>;
  try {
    stat = lstatSync(path);
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
    mkdirSync(path, { mode: opts.createMode });
    stat = lstatSync(path);
  }
  if (stat.isSymbolicLink()) throw new Error(`refusing a symlinked run-scratch path: ${path}`);
  if (!stat.isDirectory()) throw new Error(`refusing a non-directory run-scratch path: ${path}`);
  if (opts.rejectGroupWorldWrite && (stat.mode & 0o022) !== 0) {
    throw new Error(`refusing a group/world-writable run-scratch path: ${path}`);
  }
}

/**
 * Establish a run root beneath this checkout without following redirected checkout content.
 * Symlinks above `cwd` remain legal; every existing component from `.perk` through the run root
 * must be a real directory without group/world write permission, and missing components are
 * created no broader than 0755 even under a permissive umask. Run-id validation happens before
 * the first filesystem write.
 */
export function ensureRunScratch(cwd: string, runId: string): string {
  assertSafeRunId(runId);
  const dir = runScratchDir(cwd, runId);
  const components = [
    join(cwd, ".perk"),
    join(cwd, ".perk", "workflow"),
    scratchDir(cwd),
    join(scratchDir(cwd), "runs"),
    dir,
  ];
  for (const component of components) {
    ensureUnredirectedDirectory(component, {
      createMode: 0o755,
      rejectGroupWorldWrite: true,
    });
  }

  const expected = join(realpathSync(cwd), relative(cwd, dir));
  if (realpathSync(dir) !== expected) {
    throw new Error(`refusing a redirected run-scratch dir: ${dir}`);
  }
  return dir;
}

/**
 * Create the private run-owned agent directory as 0700 from the outset, then re-apply that mode on
 * reuse. The mode protects against other OS users; it is not a sandbox from another process running
 * as the same user.
 */
export function ensureAgentScratch(cwd: string, runId: string): string {
  const runDir = ensureRunScratch(cwd, runId);
  const dir = agentScratchDir(cwd, runId);
  ensureUnredirectedDirectory(dir, {
    createMode: 0o700,
    rejectGroupWorldWrite: false,
  });
  const expected = join(realpathSync(runDir), "agent");
  if (realpathSync(dir) !== expected) {
    throw new Error(`refusing a redirected agent scratch dir: ${dir}`);
  }
  chmodSync(dir, 0o700);
  return dir;
}

/**
 * The run-scoped structured run-event stream (contracts §8.12) — an NDJSON file under the
 * gitignored run scratch dir. Co-located with the run's read-only-child scratch so a runner/reader
 * finds all run artifacts under one dir.
 */
export function runEventsPath(cwd: string, runId: string): string {
  return join(runScratchDir(cwd, runId), "events.ndjson");
}

/** Names of all run scratch dirs (used to enumerate fork siblings). */
export function listRunIds(cwd: string): string[] {
  const dir = join(scratchDir(cwd), "runs");
  if (!existsSync(dir)) return [];
  return readdirSync(dir, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name);
}

// --- plan-ref: the active plan->branch ref pointer (plan-ref.json) -----------------------

/** The provider-agnostic plan ref (contracts.md §8.4); the TS twin of perk.plan.PlanRef. */
export interface PlanRef {
  provider: string;
  pr_id: string;
  url: string;
  labels: string[];
  objective_id: string | null;
  // The pinned target branch; Python-owned, parity-only on the warm plane.
  base?: string | null;
}

export function planRefPath(cwd: string): string {
  return join(workflowDir(cwd), "plan-ref.json");
}

export function readPlanRef(cwd: string): PlanRef | null {
  return readJsonOrNull<PlanRef>(planRefPath(cwd), "plan-ref");
}

export function writePlanRef(cwd: string, ref: PlanRef): void {
  mkdirSync(workflowDir(cwd), { recursive: true });
  atomicWriteFileSync(planRefPath(cwd), `${JSON.stringify(ref, null, 2)}\n`);
}

// --- hunk-watch: the watch-feedback bridge family (contracts.md §8.58) ---------------------
//
// Worktree-local, disposable: append-only NDJSON streams plus the single-consumer lease dir.
// This module is the INTERIOR construction site for the family; the hunk-plane twin is the
// self-contained bundled publisher (extension/hunkFeedback/perkFeedback.ts), pinned to these
// helpers by a path-parity test.

export function hunkWatchDir(cwd: string): string {
  return join(workflowDir(cwd), "hunk-watch");
}

/** Append-only feedback records (the Hunk publisher writes; the Pi receiver reads). */
export function hunkOutboxPath(cwd: string): string {
  return join(hunkWatchDir(cwd), "outbox.ndjson");
}

/** Append-only delivery acknowledgements (the Pi receiver writes). */
export function hunkDeliveredPath(cwd: string): string {
  return join(hunkWatchDir(cwd), "delivered.ndjson");
}

/** The single-consumer lease dir (atomic mkdir is the acquisition primitive). */
export function hunkConsumerLockDir(cwd: string): string {
  return join(hunkWatchDir(cwd), "consumer.lock");
}

export function hunkLeasePath(cwd: string): string {
  return join(hunkConsumerLockDir(cwd), "lease.json");
}

// --- markers (existence-only) ------------------------------------------------------------

/** The land->learn semaphore; the TS twin of perk.state.cache.PENDING_LEARN. */
export const PENDING_LEARN = "pending-learn";

export function markerPath(cwd: string, name: string): string {
  return join(workflowDir(cwd), "markers", name);
}

export function setMarker(cwd: string, name: string): void {
  mkdirSync(join(workflowDir(cwd), "markers"), { recursive: true });
  // Routed through the atomic seam for uniformity (empty content is trivially safe either way).
  atomicWriteFileSync(markerPath(cwd, name), "");
}

export function hasMarker(cwd: string, name: string): boolean {
  return existsSync(markerPath(cwd, name));
}

export function clearMarker(cwd: string, name: string): void {
  rmSync(markerPath(cwd, name), { force: true });
}
